codeforces support

daxida opened the pull request:
Added another script to /providers to query codeforces.com stats.

![Captura de pantalla 2023-05-24 a las 17 13 07](https://github.com/Andriamanitra/code-game-lang-stats/assets/80173028/96edbaa4-02e2-40fc-b4cf-cb205ec943f7)

From what I've been able to test the json entries of the cf API don't count multiple submissions so that's good.

Also note that the query goes over a potentially long json "from=1&count=10000". Doesn't matter for the size of my personal stats but it may do for some users?

daxida committed:
codeforces support

Andriamanitra commented on src/providers/codeforces.cr:
```diff
@@ -0,0 +1,37 @@
+      next if item["verdict"] != "OK"
+
+      language = item["programmingLanguage"].as_s
+      # Removes espaces for the JSON to work later on:
+      #     PyPy 3-64 -> PyPy-3-64
+      language = language.gsub(/ /, "-")
```
This should not be necessary. I would add `"pypy364" => "Python"` (along with other possible PyPys) to the `@@lang_aliases` in site-stats.cr instead.

daxida replied:
If spaces are not removed you get the following error:

DOMException: Failed to execute 'add' on 'DOMTokenList': The token provided ('PyPy 3-64') contains HTML space characters, which are not valid in tokens.
    at renderStatsTable (http://localhost:8080/main.js:64:23)
    at http://localhost:8080/main.js:101:13
    
I will remove the spaces then add the alias instead.

Andriamanitra commented on src/providers/codeforces.cr:
```diff
@@ -0,0 +1,37 @@
+class Codeforces < SiteStats
+  @@url_regex = Regex.new("https?://codeforces[.]com/profile/(?<username>\\w+)")
+  # https://codeforces.com/apiHelp
+  @@api_url = "https://codeforces.com/api/user.status?handle=%s&from=1&count=10000"
```
Some of the top users had over 7000 submissions already, so I think if we change this 10000 to like 50000 or something it should be safe for the foreseeable future. Other option would be to implement pagination (something similar to what the CodeWars and Exercism providers do), but I think one big request is better than multiple small ones for our purposes.

Andriamanitra commented on src/providers/codeforces.cr:
```diff
@@ -0,0 +1,37 @@
+class Codeforces < SiteStats
+  @@url_regex = Regex.new("https?://codeforces[.]com/profile/(?<username>\\w+)")
```
I would add `(www[.])?` since https://www.codeforces.com/profile/rafarafa is also a valid codeforces profile url.

Andriamanitra commented:
> From what I've been able to test the json entries of the cf API don't count multiple submissions so that's good.

Looks like it does count duplicates because it shows I have 3 solutions even though I've only solved two problems: https://codeforces.com/api/user.status?handle=Andriamanitra

It would be pretty simple to only count solutions to unique problems if we want. I'm actually not sure how the existing providers work for multiple solutions to the same problem in the same language, may need to investigate...

daxida committed:
codeforces fixes

daxida commented:
> This should not be necessary. I would add "pypy364" => "Python" (along with other possible PyPys) to the @@lang_aliases in site-stats.cr instead.

I saw what you meant, fixed.

> Some of the top users had over 7000 submissions already, so I think if we change this 10000 to like 50000

Tried 100000 with the top user and it works fine. Feel free to tune if you think it may affect performance.

> I would add (www[.])?

Fixed

> It would be pretty simple to only count solutions to unique problems if we want. 

Added a hash that keeps track of previously counted solutions for a given problem. It does count as different, solutions to the same problem but in different languages (same as codingame).

Also changed the transformed keys because it was replacing aliases of the same languages. If you had something like {gnuc17++ : 17, c++: 3, gnuc20++: 20} the previous version will only display "20" problems in c++. Now for the top user:

![Captura de pantalla 2023-05-25 a las 9 33 48](https://github.com/Andriamanitra/code-game-lang-stats/assets/80173028/adac5527-27dc-41cc-ae54-ba958b6c8b0f)

daxida committed:
removed .DS_Store

Andriamanitra committed:
Codeforces allows dashes and periods in usernames

Andriamanitra committed:
add lang_alias method, special case for C++

Andriamanitra committed:
more reliable duplicate solution checking

Andriamanitra committed:
add Codeforces to list of supported sites
